[Inductor] add expm1 lowering

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89956
* __->__ #89961

Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
[Inductor] add expm1 decomp

[ghstack-poisoned]

ngimel commented on torch/_inductor/decomposition.py:
```diff
@@ -411,6 +411,9 @@ def nan_to_num(x, nan=0.0, posinf=None, neginf=None):
     return torch.logical_not(torch.any(torch.logical_not(input)))
 
+@register_decomposition([aten.expm1.default])
+def expm1(input):
```
We need to do a lowering with a libdevice/std function (similar to erf), this is not enough accuracy. While you are at it, can you also delete log1p decomp from inductor and do the same lowering thing? With the tests for values close to one to check accuracy.

eellison replied:
I looked, and `expm1` isn't listed in the [ptx instruction set](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html). What am I missing ?

eellison replied:
nvm it's here https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_expm1.html

ngimel replied:
It's not in ptx, it's in triton's wrapper for libdevice that we call like this https://github.com/pytorch/pytorch/blob/master/torch/_inductor/codegen/triton.py#L211 for a lot of functions that are not in ptx set.

eellison committed:
Update on "[Inductor] add expm1 decomp"


Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "[Inductor] add expm1 decomp"


Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel reviewed (approved):
Awesome, thanks!

ngimel commented on test/inductor/test_torchinductor.py:
```diff
@@ -2693,6 +2693,34 @@ def fn(x):
+            )
+            self.common(
+                fn,
+                (torch.arange(-0.00001, 0.00001, 0.001).to(dtype=dtype),),
```
the step here is too large, you end up with a single value, -1e-5

eellison replied:
oops, good catch

eellison committed:
Update on "[Inductor] add expm1 decomp"


Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "[Inductor] add expm1 decomp"


Improves perf of inductor no-cudagraphs on nvidia-deeprecommender from 0.88 -> .96. I am looking into disabling implicit fallbacks for benchmark models in another pr. 


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -400,6 +400,14 @@ def sqrt(x):
         return f"1 / std::sqrt({x})"
 
+    @staticmethod
+    def log1p(x):
```
@jgong5, can these be vectorized?

eellison commented:
@pytorchbot merge
